lib/IRGen/IRGenSIL.cpp

```diff
@@ -2266,6 +2266,10 @@ Callee LoweredValue::getCallee(IRGenFunction &IGF,
   switch (kind) {
   case Kind::FunctionPointer: {
     auto &fn = getFunctionPointer();
+    if (calleeInfo.OrigFnType->getRepresentation() ==
+        SILFunctionTypeRepresentation::ObjCMethod) {
+      return getObjCDirectMethodCallee(std::move(calleeInfo), fn, selfValue);
+    }
```

Comment:
We have to do something extra on the caller side when making a direct call of a class method, right?

Reply:
What else is needed on the caller side? I have objc_direct class method test cases (`directClassMethod`, `directClassMethod2`) that seem to be working.

Reply:
Well, in general you have to make sure the class is realized before you can call direct class methods.  But in fact it looks like Swift does this explicitly before calling *any* class method, which is unfortunate — but it means there's nothing to do here except make sure we test that pattern, so that if Swift ever gets smarter here we can avoid some work.

Reply:
I think this makes sense, but I don't know how to create this test. Would it be something like calling a class method that has an interface but no implementation?

Reply:
You can just do test the IRGen output for a call to a direct class method that looks something like `TheClass.directClassMethod()`.

Reply:
Maybe I'm misunderstanding something but I'm already testing this line in IRGen, `markUsed(Bar.directClassMethod())`.
https://github.com/apple/swift/pull/33349/files#diff-836f9a570b4e9833ab0598773438d6dbR39

Reply:
You're not testing that we force the class to be realized before we make the call.

Reply:
Ok, but I don't understand how to realize a class. Still have a lot to learn about Swift/Objective-C

Reply:
I just mean that you should have an IRGen test that checks for the exact sequence of materialization for the class argument, and you should leave a comment saying that we need to continue to do this for direct calls even if we stop doing it normally.

Reply:
Ah thanks, I finally understand! I've added more checks here:
https://github.com/apple/swift/pull/33349/files#diff-5115fb955c7cba0a551223ce515367063ba8f83c939474d54515f4477bd4efe4R39-R51

Reply:
Thanks, that's exactly what I wanted.